### PR TITLE
feat:adding a web assembly playground for the examples

### PR DESCRIPTION
**Pull request by Dominik1999**
Updated the Miden Assembly Playground from [here](https://github.com/timgestson/miden-assembly-playground) to accept inputs (`stack_init` and `advice_tape`) and to work with all our initial examples.

### CODE REVIEW

File: miden-wasm/src/lib.rs
Comment (tohrnii) [inline miden-wasm/src/lib.rs:14]: nit: It would be good to have doc comment for the method.

File: miden-wasm/src/lib.rs
Comment (tohrnii) [inline miden-wasm/src/lib.rs:26]: nit: It would be good to have doc comment for the method.

File: miden-wasm/src/lib.rs
Comment (tohrnii) [inline miden-wasm/src/lib.rs:19]: minor nit (feel free to ignore): I think using `input` instead of `inputs_des` might be better.

File: miden-wasm/src/lib.rs
Comment (tohrnii) [inline miden-wasm/src/lib.rs:63]: Maybe these lines can be removed for now?

File: package.json
Comment (tohrnii) [inline package.json:12]: Is this import required?

File: .gitignore
Comment (tohrnii) [inline .gitignore:6]: Are these required?

File: miden-wasm/src/lib.rs
Comment (tohrnii) [inline miden-wasm/src/lib.rs:41]: minor nit: Can this line be removed?

File: src/App.tsx
Comment (tohrnii) [inline src/App.tsx:84]: We could return an object with `inputs` and `code` instead of calling the function twice.

File: src/App.tsx
Comment (tohrnii) [inline src/App.tsx:31]: We should probably return an object with inputs and code instead of calling the function twice and taking inputs boolean flag.

File: src/App.tsx
Comment (tohrnii) [inline src/App.tsx:7]: nit: The linter complains that this import is unused so maybe we should remove it.

**Review by tohrnii — CHANGES_REQUESTED** [review]
Thank you @Dominik1999, looks good! I left a few minor comments inline. This is not a full review yet. I still need to figure out a few things.

Also, we should probably not use both `yarn` and `npm`. We should stick to one and remove the other lock file. I don't understand the nuances well enough to have an opinion so we can go with whichever of the two you think is best.

File: src/App.tsx
Comment (tohrnii) [inline src/App.tsx:65]: The nPrime example seems to be missing from the examples directory. We should either add it in the examples or remove the entry from here.

File: README.md
Comment (bobbinth) [inline README.md:33]: I would put this into a separate section towards the end - maybe "Acknowledgement" or something like this?

File: README.md
Comment (bobbinth) [inline README.md:13]: Maybe put this under a common section - something like "Available examples"?

File: README.md
Comment (bobbinth) [inline README.md:43]: This feels like it should be a separate section which describes how to run the site locally - so, maybe put this under "Running the playground locally" section?

**Review by bobbinth — COMMENTED** [review]
I did a quick read (mainly through the readme file) and left a couple of small comments inline.

Also, we should probably add a License section at the end (similar to how we do it in the main Miden repo).

Another comment for the future: I would probably prefer using [Deno](https://github.com/denoland/deno) over Node.js. I believe (though could be wrong), it would make the repo cleaner and Deno also has very good integration with Rust.

File: .gitignore
Comment (Dominik1999) [reply .gitignore:1]: done

File: README.md
Comment (Dominik1999) [reply README.md:9]: done

File: miden-wasm/src/lib.rs
Comment (Dominik1999) [reply miden-wasm/src/lib.rs:14]: What is a doc comment?

File: miden-wasm/src/lib.rs
Comment (Dominik1999) [reply miden-wasm/src/lib.rs:26]: what is a doc comment?

File: miden-wasm/src/lib.rs
Comment (Dominik1999) [reply miden-wasm/src/lib.rs:63]: yes indeed. Would be cool to be able to prove later though

File: miden-wasm/src/lib.rs
Comment (Dominik1999) [reply miden-wasm/src/lib.rs:41]: indeed. Was a leftover

File: src/App.tsx
Comment (Dominik1999) [reply src/App.tsx:7]: Hmm, but we use all of them but `FormGroup`. I removed that now

Comment (tohrnii) [reply src/App.tsx]: @Dominik1999 Ah sorry, I should have clarified. Github automatically marks additional 3 lines above (probably to make locating the code easier). I did mean only `FormGroup`.

File: miden-wasm/src/lib.rs
Comment (grjte) [reply miden-wasm/src/lib.rs:14]: https://doc.rust-lang.org/rust-by-example/meta/doc.html

File: src/App.tsx
Comment (grjte) [inline src/App.tsx:113]: We can improve this `Select` by following the MaterialUI `Select` component pattern. I also think we should move the outputs and execute button to the top and make a few minor adjustments to the styling. I propose:

1. Delete everything below the source code `CodeMirror`
2. Replace this highlighted box with the box code below.
3. Replace `handleChangeMultiple` with `handleSelectChange` code below.
4. Change the type of `example` and the input to `getExample` to `string` instead of `string[]`

```
<Box sx={{ display: 'flex', flexDirection: 'row', mx: 'auto', my: 4}}>
          <FormControl variant="outlined" sx={{minWidth: 120}}>
              <InputLabel shrink htmlFor="select-example">
                Example
              </InputLabel>
              <Select
                value={example}
                onChange={handleSelectChange}
                labelId='select-example'
              >
                {examples.map((name) => (
                  <MenuItem value={name} key={name}>{name}</MenuItem>
                ))}
              </Select>
          </FormControl>
          <FormControl variant="outlined" sx={{minWidth: 120, mx: 2}}>
              <InputLabel shrink htmlFor="select-outputs">
                Outputs
              </InputLabel>
              <Select
              value={numOfOutputs}
              onChange={(e) => setNumOfOutputs(Number(e.target.value))}
              labelId='select-outputs'
            >
              <MenuItem value={0} key={0}>0</MenuItem>
              <MenuItem value={1} key={1}>1</MenuItem>
              <MenuItem value={16} key={16}>16</MenuItem>
            </Select>
          </FormControl>
          <Button
            variant="contained"
            color="primary"
            onClick={() => {
              init().then(() => {
                try {
                  let resp = program(code, inputs, numOfOutputs);
                  console.log(resp);
                  setOutput({
                    init: true,
                    success: true,
                    text: resp.join(" "),
                  });
                } catch (error) {
                  setOutput({
                    init: true,
                    success: false,
                    text: "Error: Check the developer console for details.",
                  });
                }
              });
            }}
            endIcon={<SendIcon />}
          >
            Execute
          </Button>
          {output.init ? (
            <Alert
              variant="outlined"
              severity={output.success ? "success" : "error"}
              sx={{mx: 2}}
            >
              <Typography color="inherit" component="div">
                {output.text}
              </Typography>
            </Alert>
          ) : (
            <></>
          )}
        </Box>
```

```
  const handleSelectChange = async (event: SelectChangeEvent) => {
    const value = event.target.value;
    // set the current example to the selected one
    setExample(value)

    // retrieve the example data and update the inputs and code
    const example_code = await getExample(value)
    setInputs(await example_code[0])
    setCode(await example_code[1])    

    // reset the output
    setOutput(emptyOutput)
  }
```

File: src/App.tsx
Comment (grjte) [inline src/App.tsx:48]: let's add a constant for empty output here so we can reset it when the example is changed:

```

  const emptyOutput = {
    init: false,
    success: true,
    text: "",
  };
  const [output, setOutput] = React.useState(emptyOutput);
```

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @Dominik1999, this is really cool!

I left comments inline about, including some proposed changes to tidy up the ui/ux. Let me know if you have any questions!

Also, not all of the examples are working for me. I'm getting a few 404s. Can you check on this?

File: src/App.tsx
Comment (grjte) [reply src/App.tsx:113]: tbh, it actually probably makes sense to define this as its own component and then put it at both the top and the bottom for better UX on the long examples

File: README.md
Comment (Dominik1999) [reply README.md:9]: Got it, one question, though. [Here](https://maticnetwork.github.io/miden/intro/overview.html#inputs-and-outputs) we say we can have two type of external inputs and I think the advice provider can entail three different kinds of secret inputs, right?

File: README.md
Comment (Dominik1999) [reply README.md:50]: good point

File: miden-wasm/Cargo.toml
Comment (Dominik1999) [reply miden-wasm/Cargo.toml:11]: I think it makes most sense to have the playground on the latest stable to be honest

File: src/App.tsx
Comment (Dominik1999) [reply src/App.tsx:13]: I will open an issue that this will be the first refactoring

File: src/App.tsx
Comment (Dominik1999) [reply src/App.tsx:14]: I will add this to the refactoring issue

Comment (Dominik1999) [reply src/App.tsx:14]: https://github.com/0xPolygonMiden/examples/issues/7

File: src/App.tsx
Comment (Dominik1999) [reply src/App.tsx:19]: Jep we can delete that

File: src/theme.ts
Comment (Dominik1999) [reply src/theme.ts:17]: How nice of you that you already gave me the right color codes here!

**Comment by Dominik1999** [comment]
Thank so much for your review @grjte. The two non-working examples should work when we rabase and merge.

**Review by Dominik1999 — COMMENTED** [review]
Thanks for the review!

File: README.md
Comment (bobbinth) [inline README.md:35]: The title is repeated here. Also, I would use first level title (i.e., `#`) only at the very top (for "Miden Assembly Playground") and would "downgrade" all other titles. For example, here I'd use second-level title (i.e., `##`).

**Comment by bobbinth** [comment]
One other comment: I wonder whether rather than putting the playground into the root of the repo it should go into something like `/wasm-playground` (or just `/playground`) directory. This would enable us to add other things in the future (e.g., CLI mentioned in #9). Maybe the structure of the repo could be something like:

```
├─ bench-cli
├─ examples
└─ playground
```

What do you guys think?

**Comment by Dominik1999** [comment]
Thanks @bobbinth ! 

I think it makes sense. Maybe we can merge this PR as is and I open a issue to change the structure and add the CLI?

File: src/App.tsx
Comment (grjte) [inline src/App.tsx:173]: As suggested by @tohrnii, let's set a max height here. I think 500px is ok. We'll revisit later when upgrading the design anyway

```
maxHeight="500px"
```

File: src/App.tsx
Comment (grjte) [inline src/App.tsx:164]: Let's set a max height here. I think 200px is probably ok for this one.

File: src/App.tsx
Comment (grjte) [inline src/App.tsx:159]: For the future: It would be good to pull this out into its own component and add it at both the top and the bottom of the screen. I think that would provide better UX (people wouldn't need to scroll back up after writing their code). Do you agree? If so, can you create an issue?

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @Dominik1999! I have a few more small things that could use adjustment, as well as a failure when I run `wasm-pack test --node`

**Comment by grjte** [comment]
> One other comment: I wonder whether rather than putting the playground into the root of the repo it should go into something like `/wasm-playground` (or just `/playground`) directory.

I agree with the structure suggestion too, but I think it's ok to merge this once the small points are addressed so there's something to show for the discussion today. Please update here with the issue link for restructure + CLI once you've created it @Dominik1999

File: src/App.tsx
Comment (Dominik1999) [reply src/App.tsx:173]: good point, somehow that slipped

File: src/App.tsx
Comment (Dominik1999) [reply src/App.tsx:159]: Yes, I will create an issue

**Comment by Dominik1999** [comment]
Thanks @grjte 

here is the issue for changing the repo structure: https://github.com/0xPolygonMiden/examples/issues/11

File: README.md
Comment (grjte) [inline README.md:34]: nit: `nPrime` is missing from here. We could add something like:

"nPrime: A program which calculates the nth prime number, using the VM memory."

**Review by grjte — APPROVED** [review]
Looks good, thank you! I left one final comment that would be good to update before merging, but I'm approving so you're not blocked.

Can you please make sure that everything we discussed creating issues for gets created? (e.g. there's a TODO in the code that doesn't seem to be tracked by an issue yet, unless I just missed the issue when I looked?)

File: README.md
Comment (Dominik1999) [reply README.md:34]: it is there now, thanks again!

**Review by grjte — APPROVED** [review]

**Review by tohrnii — APPROVED** [review]
All looks good! Thank you!

**Pull request merged**
